### DISCUSSION TITLE
Diff from `magit-diff-paths` is without colors while other Magit diffs are not

### DISCUSSION THREAD

**Discussion by gustav-b** [magit/magit#5093] [Support] [answered]
Using `Magit 20240211.1712 [>= 3.3.0.50-git], Transient 0.5.3, Git 2.43.2, Emacs 29.2, darwin`

Not sure if this is expected, but if I diff two files in a working directory using `d p` (`magit-diff-paths`) the resulting buffer is without any colors/highlighting. Any other diff I get from Magit is colored. For example:

`magit-diff-working-tree`:
<img width="1040" alt="magit-diff-working-tree" src="https://github.com/magit/magit/assets/368219/af17611f-9774-422c-aafb-03c673f1cc5b">

`magit-diff-paths`:
<img width="1037" alt="magit-diff-paths" src="https://github.com/magit/magit/assets/368219/f2dd9286-48d7-4d84-82a9-8f5bde557b70">

Is this a bug? If not, how do I enable coloring for path diffs?

**Comment by kyleam** [comment]
> Is this a bug?

Yes.

As of 78a979fd (Support displaying errors that occur when washing a section, 2023-03-18), the diff buffer is washed only if the `git diff` command exits with a zero status.  That's a problem with `git diff --no-index ...` because exits with 1 if there are differences (like `diff(1)`).

**Comment by tarsius** [comment, marked as answer]
Now fixed with 0e8f25a8d8011328f2bf082232c720b24c2a12c2.

**Comment by gustav-b** [comment]
It seems that this issue is back again. 😕 

This is the same test case with `Magit 20250327.1533 [>= 20250327.1533], Transient 20250324.47, Git 2.49.0, Emacs 30.1, darwin`

<img width="800" alt="magit-diff-2025-03-28" src="https://github.com/user-attachments/assets/baa11345-844b-4ff0-8782-56557d32e50d" />

**Reply by kyleam** [reply]
Bisects to 1f17e515 (magit--git-{insert,wash}: Potentially use complete error text, 2024-11-15).

**Reply by kyleam** [reply]
I haven't had a time to analyze the wider context and need to head out the door , but fwiw perhaps it's just a matter of restricting that `full` fallthrough to `t` values so that `wash-anyway` doesn't trigger it (though, even if that's correct, @tarsius may want to fix this in another way or elsewhere):

```diff
diff --git a/lisp/magit-git.el b/lisp/magit-git.el
index 19f02f9a..807fcb81 100644
--- a/lisp/magit-git.el
+++ b/lisp/magit-git.el
@@ -548,7 +548,9 @@ (defun magit--git-wash (washer keep-error &rest args)
   (declare (indent 2))
   (setq args (flatten-tree args))
   (let ((beg (point))
-        (exit (magit--git-insert (and keep-error 'full) args)))
+        (exit (magit--git-insert
+               (if (eq keep-error t) 'full keep-error)
+               args)))
     (when (stringp exit)
       (goto-char beg)
       (insert (propertize exit 'face 'error))
```

**Reply by tarsius** [reply]
Fixed again, this time with 64e69ea809042a1182c7a0b3e92c1f7c6b4f8b6b.
